server/src/main/scala/org/http4s/server/package.scala

```diff
@@ -3,12 +3,47 @@ package org.http4s
 import scalaz.concurrent.Task
 
 package object server {
-  type Middleware  = HttpService => HttpService
-  type HttpService = Service[Request, Response]
+  type Service[-A, +B] = A => Task[B]
+
+  type PartialService[-A, +B] = Service[A, Option[B]]
 
+  type HttpService = Service[Request, Response]
   object HttpService {
-    def apply(pf: PartialFunction[Request, Task[Response]]): HttpService = Service(pf)
+    /**
+     * Lifts a partial function to a service.  The resulting service will
+     * return a Task containing `None` where the partial function is not
+     * defined, and Some result where it is.
+     */
+    def apply(pf: PartialFunction[Request, Task[Response]]): HttpService =
+      pf.lift.andThen {
+        case Some(respTask) => respTask
+        case None => Task.now(Response(Status.NotFound))
+      }
+
+    val empty: HttpService = Function.const(Task.now(Response(Status.NotFound)))
   }
+
+  implicit class HttpServiceSyntax(val service: HttpService) extends AnyVal {
+    def orElse(that: HttpService): HttpService = { req: Request =>
+      service(req).flatMap {
+        case resp: Response if resp.status == Status.NotFound =>
+          that(req)
+        case resp =>
+          Task.now(resp)
+      }
+   }
+  }
+
+  type Middleware[A, B, C, D] = Service[A, B] => Service[C, D]
+  object Middleware {
+    def apply[A, B, C, D](f: (C, Service[A, B]) => Task[D]): Middleware[A, B, C, D] = {
```

Comment:
The `orElse` part? I currently use it quite a bit, but I cant speak for anybody but myself. To be honest, I've started using rho in my personal projects more and it will soon (hopefully) get its own `orElse` like behavior so I doubt I'll be needing it in the long term.